server/src/main/java/org/elasticsearch/ReleaseVersions.java

```diff
@@ -0,0 +1,128 @@
+
+    static NavigableMap<Integer, String> readVersionsFile(Class<?> versionContainer) {
+        String versionsFileName = versionContainer.getSimpleName() + ".csv";
+        URL versionsFile = versionContainer.getResource(versionsFileName);
+        if (versionsFile == null) {
+            Log.error("Could not find versions file for class [{}]", versionContainer);
+            return Collections.emptyNavigableMap();
+        }
+
+        try (
+            BufferedReader reader = new BufferedReader(
+                new InputStreamReader(versionContainer.getResourceAsStream(versionsFileName), StandardCharsets.UTF_8)
+            )
+        ) {
+            NavigableMap<Integer, String> versions = new TreeMap<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (COMMENT_LINE.matcher(line).find()) continue; // a comment
+
+                var matcher = VERSION_LINE.matcher(line);
+                if (matcher.matches() == false) {
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile);
+                    continue;
+                }
+                try {
+                    Integer id = Integer.valueOf(matcher.group(2));
+                    String version = matcher.group(1);
+                    String existing = versions.putIfAbsent(id, version);
+                    if (existing != null) {
+                        Log.warn("Duplicate id [{}] for versions [{}, {}] in [{}]", id, existing, version, versionsFile);
+                    }
+                } catch (NumberFormatException e) {
+                    // cannot happen??? regex is wrong...
+                    assert false : "Regex allowed non-integer id through: " + e;
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile, e);
+                }
+            }
+
+            return Collections.unmodifiableNavigableMap(versions);
+        } catch (Exception e) {
+            Log.error("Could not read versions file [{}]", versionsFile, e);
+            return Collections.emptyNavigableMap();
+        }
+    }
+
+    private static NavigableMap<Integer, String> findVersionsMap(Class<?> versionContainer) {
+        return RESOLVED_VERSIONS.computeIfAbsent(versionContainer, ReleaseVersions::readVersionsFile);
+    }
+
+    public static String findReleaseVersion(Class<?> versionContainer, int id) {
+        if (USES_VERSIONS == false) return Integer.toString(id);
+
+        NavigableMap<Integer, String> versions = findVersionsMap(versionContainer);
+        String exactVersion = versions.get(id);
+        if (exactVersion != null) {
+            return exactVersion;
+        }
+
+        // go for a range
+        if (versions.isEmpty()) {
+            // uh oh, something went wrong reading the file
+            return "<" + id + ">";
+        }
+
+        StringBuilder range = new StringBuilder();
+
+        var lowerRange = versions.lowerEntry(id);
```

Comment:
If we fall down to here, something is wrong. On prem users would only ever be using released versions. There shouldn't be any uses of eg transport versions between on-prem releases. I think we could simplify this and fallback to just the id as above in the "the version map is empty" case (which also should never happen...maybe let's add an assert for this case?)

Reply:
I'm largely thinking of snapshot builds where we don't have a release version, but its a valid transport/index version. We could add an assert for the lower range, but we still need to do something if this ever happens, somehow, in production. We shouldn't blow up because an error message can't be resolved properly.

Reply:
I've added an assert in the rewritten code